Improve error handling of experimental features in non-experimental mode

yongtang opened the pull request:
**- What I did**

This fix tries to address several issues raised in #28626 where run against a non-experimental daemon may not generate correct error message:
1. Incorrect flags were not checked against the supported features:
   ```
   $ docker stack --nonsense
   unknown flag: --nonsense
   ```
2. Subcommands were not checked against the supported features:
   ```
   $ docker stack ls
   Error response from daemon: This node is not a swarm manager...
   ```

**- How I did it**

This fix address the above mentioned issues by:
1. Add a pre-check for FlagErrorFunc
2. Recursively check if a feature is supported for cmd and its parents.

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

![misfondoshd com_hermoso-gatito-recostado_1366x768](https://cloud.githubusercontent.com/assets/6932348/20465076/0fb5d6cc-af0a-11e6-94eb-518f42804a4b.jpg)


This fix related to #28626. There is on remaining issue in #28626 where a feature was not checked to see if it is supported:
```
$ docker deploy
"docker deploy" requires exactly 1 argument(s).
```

This one is handled in `Args`. Will investigate further to see if it is possible to fix this specific issue as well.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Improve error handling of experimental features in non-experimental mode

This fix tries to address several issues raised in 28626 where
run against a non-experimental daemon may not generate correct
error message:
1. Incorrect flags were not checked against the supported features:
   ```
   $ docker stack --nonsense
   unknown flag: --nonsense
   ```
2. Subcommands were not checked against the supported features:
   ```
   $ docker stack ls
   Error response from daemon: This node is not a swarm manager...
   ```

This fix address the above mentioned issues by:
1. Add a pre-check for FlagErrorFunc
2. Recursively check if a feature is supported for cmd and its parents.

This fix fixes 28626.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
cc @albers @thaJeztah @vieux @mlaventure

yongtang commented:
There seems to be some issues with the interaction between daemon and client. Let me close this PR for now.

Will reopen if I could fix the issue.

yongtang commented:
@albers @thaJeztah @vieux @mlaventure The issue has been fixed. Please take a look.

albers commented:
Thanks @yongtang. I'm sorry, but there are still invocations that produce an error message where I would expect them not to do so:
- `docker stack hugo`
- `docker checkpoint create`
- `docker checkpoint hugo`
- `docker deploy`

yongtang committed:
Additional experimental features in non-experimental mode error handling

This fix is the follow up of the last commit.
In this fix:
1. If any of the parents of a command has tags, then this command's
`Args` (Args validation func) will be wrapped up. The warpped up func
will check to see if the feature is supported or not. If it is not supported,
then a not supported message is generated instead.

This fix is related to 28626.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @albers. The PR has been updated. Please take a look and let me know if there are any issues.

albers commented:
Seems `docker stack` is no longer experimental, so the command now behaves like any non-experimental command. => Everything OK with this command in this PR.
With the other commands: I found no issues, everything works as expected. Thanks, @yongtang!
LGTM

LK4D4 commented on cmd/docker/docker.go:
```diff
@@ -82,9 +99,53 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
+// This is different from the VisitAll of cobra.Command where only parents
+// are checked.
+func visitAll(root *cobra.Command, fn func(*cobra.Command)) {
+	var traverse func(r *cobra.Command, fn func(*cobra.Command))
```
this looks extremely weird to me. Why you need closure at all?

yongtang replied:
Thanks @LK4D4 for the review. The PR has been updated and `visitAll` has been simplified. Please take a look and let me know if there are any issues.

LK4D4 commented:
@yongtang posted comment

yongtang commented:
Ping @LK4D4 @thaJeztah please take a look.

aaronlehmann commented on cmd/docker/docker.go:
```diff
@@ -52,6 +52,23 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
+	// is called.
+	flagErrorFunc := cmd.FlagErrorFunc()
+	cmd.SetFlagErrorFunc(func(cmd *cobra.Command, err error) error {
+		if dockerCli.Client() == nil { // when using --help, PersistenPreRun is not called, so initialization is needed.
```
typo in comment

aaronlehmann commented on cmd/docker/docker.go:
```diff
@@ -52,6 +52,23 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
+	})
 
 	cmd.SetHelpFunc(func(ccmd *cobra.Command, args []string) {
 		if dockerCli.Client() == nil { // when using --help, PersistenPreRun is not called, so initialization is needed.
```
typo in comment

aaronlehmann commented on cmd/docker/docker.go:
```diff
@@ -82,9 +99,47 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
+		cmdArgs := ccmd.Args
+		if cmdArgs != nil {
+			ccmd.Args = func(cmd *cobra.Command, args []string) error {
+				if dockerCli.Client() == nil { // when using --help, PersistenPreRun is not called, so initialization is needed.
```
typo in comment

yongtang commented:
Thanks @aaronlehmann for the review. The PR has been updated.

thaJeztah commented:
Both Janky and Experimental timed out on the same test, not sure if it's related or something else is causing this?

https://jenkins.dockerproject.org/job/Docker-PRs/38016/console

```
02:17:51 --- FAIL: Test (3448.65s)
02:17:51 panic: DockerTrustSuite.TestBuildContextDirIsSymlink test timed out after 5m0s [recovered]
02:17:51 	panic: DockerTrustSuite.TestBuildContextDirIsSymlink test timed out after 5m0s
02:17:51 
```

https://jenkins.dockerproject.org/job/Docker-PRs-experimental/29423/console

```
02:14:06 PASS: docker_cli_swarm_test.go:831: DockerSwarmSuite.TestUnlockEngineAndUnlockedSwarm	1.569s
02:19:06 --- FAIL: Test (3463.01s)
02:19:06 panic: DockerTrustSuite.TestBuildContextDirIsSymlink test timed out after 5m0s [recovered]
02:19:06 	panic: DockerTrustSuite.TestBuildContextDirIsSymlink test timed out after 5m0s
02:19:06 
```

thaJeztah commented:
Hm, seen that same failure on another PR this morning, so should not be related; wondering what's causing it though

yongtang commented:
@thaJeztah The PR was rebased to fix the merging conflict. Now all tests passed. Please take a look.

AkihiroSuda reviewed (commented):
nit: now `newDockerCommand` is hard to follow because it has 110 lines.

Can you please split the function?

yongtang commented:
Thanks @AkihiroSuda. The PR has been updated.

LK4D4 commented on cmd/docker/docker.go:
```diff
@@ -166,9 +234,12 @@ func hideUnsupportedFeatures(cmd *cobra.Command, clientVersion string, hasExperi
-		if _, ok := cmd.Tags["experimental"]; ok {
-			return errors.New("only supported with experimental daemon")
+	// We check recursively so that, e.g., `docker stack ls` will return the same output as `docker stack`
+	for curr := cmd; curr != nil; curr = curr.Parent() {
```
can't you run this loop only if `!hasExperimental`?

yongtang replied:
Thanks @LK4D4. Done.

LK4D4 commented on cmd/docker/docker.go:
```diff
@@ -72,17 +111,46 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
-	cmd.AddCommand(newDaemonCommand())
-	commands.AddCommands(cmd, dockerCli)
+		cmdArgs := ccmd.Args
+		if cmdArgs != nil {
```
I'm not sure what `cmdArgs == nil` means, but doesn't it make sense to wrap it anyway and just return nil error?

yongtang replied:
@LK4D4 I changed the implementation so that if ccmd.Args (which is the func to check for number of Args) is nil, it will return immediately and remain as nil func.
This will reduce unnecessary scope in the previous implementation.

LK4D4 commented:
Left some minor comments.

yongtang commented:
Thanks @LK4D4 for the review. The PR has been updated. Please take a look.

vdemeester commented:
/cc @dnephin

dnephin commented on cmd/docker/docker.go:
```diff
@@ -53,8 +53,47 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
+			opts.Common.SetDefaultOptions(flags)
+			dockerPreRun(opts)
+			dockerCli.Initialize(opts)
+		}
```
This block is now in 3 places. Can this be broken out into a new function that initializes the client?

yongtang replied:
Thanks @dnephin. Done.

dnephin commented on cmd/docker/docker.go:
```diff
@@ -72,17 +111,48 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
-	commands.AddCommands(cmd, dockerCli)
+		if ccmd.Args == nil {
+			return
+		}
```
The indentation looks off in github, some missing tabs?

yongtang replied:
@dnephin The indentation seems to be OK. The issue is that diff is not able to find the logic `diff` piece:
```udiff
                        return
@@ -72,17 +100,52 @@ func newDockerCommand(dockerCli *command.DockerCli) *cobra.Command {
                        ccmd.Println(err)
                }
        })
+}
 
-       flags = cmd.Flags()
-       flags.BoolVarP(&opts.Version, "version", "v", false, "Print version information and quit")
-       flags.StringVar(&opts.ConfigDir, "config", cliconfig.Dir(), "Location of client config files")
-       opts.Common.InstallFlags(flags)
+func setValidateArgs(dockerCli *command.DockerCli, cmd *cobra.Command, flags *pflag.FlagSet, opts *cliflags.ClientOptions) {
+       // The Args is handled by ValidateArgs in cobra, which does not allows a pre-hook.
+       // As a result, here we replace the existing Args validation func to a wrapper,
+       // where the wrapper will check to see if the feature is supported or not.
+       // The Args validation error will only be returned if the feature is supported.
+       visitAll(cmd, func(ccmd *cobra.Command) {
+               // if there is no tags for a command or any of its parent,
+               // there is no need to wrap the Args validation.
+               if !hasTags(ccmd) {
+                       return
+               }
 
-       cmd.SetOutput(dockerCli.Out())
-       cmd.AddCommand(newDaemonCommand())
-       commands.AddCommands(cmd, dockerCli)
+               if ccmd.Args == nil {
+                       return
+               }
 
-       return cmd
+               cmdArgs := ccmd.Args
```

yongtang commented:
@dnephin The PR has been updated with function consolidated. Please take a look.

dnephin reviewed (approved):
LGTM

LK4D4 commented:
@yongtang needs rebase :/ sorry

yongtang commented:
Thanks @LK4D4. The PR has been rebased. Please take a look.

LK4D4 reviewed (approved):
LGTM
